dialog/dialog/relational-algebra.md

```diff
@@ -0,0 +1,415 @@
+
+A n-ary relation contains n-tuples, which can each be written:
+
+```
+(a1: v1, a2: v2, ..., an: vn)
+```
+
+Where `a1, a2, ..., an` give the names of each attribute, and `v1, v2, ..., vn` their values.
+
+# 3. Operations
+
+An operator specifies an operation against a stream, and is represented by a node. Operators MAY be stateful, and linear operators are those which can be computed using only the deltas at the current timestamp.
+
+All operators take some number of relations as input, transforming them into an output relation.
+
+| Name                                       | Arity |
+| ------------------------------------------ | ----- |
+| [Projection](#31-projection)               | 1     |
+| [Rename](#32-rename)                       | 1     |
+| [Selection](#33-selection)                 | 1     |
+| [Union](#34-union)                         | N-ary |
+| [Difference](#35-difference)               | 1     |
+| [Cartesian Product](#36-cartesian-product) | 2     |
+| [Natural Join](#37-natural-join)           | 2     |
+| [Theta Join](#38-theta-join)               | 2     |
+| [Equijoin](#39-equijoin)                   | 2     |
+| [Semijoin](#310-semijoin)                  | 2     |
+| [Antijoin](#311-antijoin)                  | 2     |
+| [Group By](#312-group-by)                  | 1     |
+
+## 3.1 Projection
+
+A projection is a unary operation which is parameterized over a set of attribute names, and restricts the tuples in its input relation to the attributes given by these names.
+
+For example:
+
+```
+projection([:id, :name], [
+    (id: 1, name: "Kōbō Abe", born: 1924),
+    (id: 2, name: "Jorge Luis Borges", born: 1899),
+    (id: 3, name: "Italo Calvino", born: 1923),
+    (id: 4, name: "Umberto Eco", born: 1932)
+]) => [
+    (id: 1, name: "Kōbō Abe"),
+    (id: 2, name: "Jorge Luis Borges"),
+    (id: 3, name: "Italo Calvino"),
+    (id: 4, name: "Umberto Eco")
+]
+```
+
+## 3.2 Rename
+
+A rename is a unary operation which is parameterized over a pair of attribute names, `(from, to)`, whose input relation contains an attribute named `from`, but does not contain an attribute named `to`. It returns a relation whose tuples are identical to those in the input, except the attribute given by `from` is renamed to `to`.
+
+For performance reasons, implementations are RECOMMENDED to support fusing such operations together, by allowing multiple attributes to be renamed using a single operation.
+
+For example:
+
+```
+rename([foo: :bar], [
+    (foo: 1),
+    (foo: 2),
+    (foo: 3)
+]) => [
+    (bar: 1),
+    (bar: 2),
+    (bar: 3)
+]
+```
+
+## 3.3 Selection
+
+A selection is a unary operation which is parameterized over a [propositional formula](#4-propositional-formula), and that returns the tuples in its input relation for which this formula holds.
+
+For example:
+
+TODO: how should I write propositional formulas?
```

Comment:
Cool, thanks! When I do a polish pass through things at the end, I might just wrap the formulas in a lambda over the tuple, to more clearly delineate both contexts. But it's good to know that things are clear as is!

Reply:
I specifically like the choice to *not* have a lambda here. Since that'd make me think that "any" computation is allowed, while only propositional formulas are.